### PR TITLE
Fix various

### PR DESCRIPTION
**Pull request by dkochmanski**
atomic fixes for a few issues reported lately. also a cosmetic improvement to slim library.

### CODE REVIEW

File: Core/clim-core/frames.lisp [hunk 1]
```diff
@@ -773,45 +768,44 @@ documentation produced by presentations.")
...
-                                           (t nil))
-                                       ,@(rest layout)
-                                       ,@(when pointer-documentation
-                                           '(%pointer-documentation%)))))
-                                 layouts)
-                         layouts)))))
-       ;; Update frame-current-panes and the special pane slots.
-       (update-frame-pane-lists frame))))
+     (with-look-and-feel-realization (fm frame)
+       (unless (frame-panes-for-layout frame)
```
Comment (scymtym) [inline Core/clim-core/frames.lisp]: I thought about this change as well but should `with-look-and-feel-realization` actually bind `*application-frame*`?

File: Core/clim-core/frames.lisp [hunk 2]
```diff
@@ -879,6 +873,24 @@ documentation produced by presentations.")
...
+
+(defmethod generate-panes (fm (frame application-frame))
+  (with-look-and-feel-realization (fm frame)
+    (unless (frame-panes-for-layout frame)
+      (setf (frame-panes-for-layout frame)
+            `((single-pane . ,(make-clim-interactor-pane :name 'single-pane)))))
+    (let ((single-pane
+            (alexandria:assoc-value (frame-panes-for-layout frame)
+                                    'single-pane :test #'eq)))
+      (setf (frame-panes frame) single-pane)))
```
Comment (scymtym) [inline Core/clim-core/frames.lisp]: Maybe add a comment explaining that `make-clim-interactor-pane` may wrap the `single-pane` instance in other panes and therefore the lookup in `frame-panes-for-layout` is needed.

File: Core/clim-core/frames.lisp (hunk 1 above)
Comment (dkochmanski) [reply Core/clim-core/frames.lisp]: It is not said in the spec, but I believe that this is the right thing to do. Otherwise frame argument would be unnecessary. Moreover from more tangible perspective, it is important that *application-frame* is bound to the current frame if we want to be able to use i.e the instance slots in pane definition. Consider the following:

```common-lisp
(define-application-frame poem-viewer ()
  ((title  :initarg :title  :reader poem-title)
   (verses :initarg :verses :reader poem-verses))
  (:panes (app :application
               :display-function (lambda (frame pane)
                                   (format pane (poem-verses frame)))))
  (:layouts
   (default
    (labelling (:label (poem-title *application-frame*)) ; <----
      app))))

(run-frame-top-level
 (make-application-frame 'poem-viewer
                         :title "CLIM"
                         :verses "Roses are red,~
                                  Violets are blue,~
                                  CLIM is software,~
                                  Which does kaboom."))
```

File: Core/clim-core/panes.lisp [hunk 3]
```diff
@@ -2584,14 +2584,18 @@ SCROLLER-PANE appear on the ergonomic left hand side, or leave set to
...
+      (cond ((consp display-function)
+	     (apply (car display-function)
+		    frame pane (cdr display-function)))
+	    (display-function
+	     (funcall display-function frame pane))
+	    (t nil))))
+  ;; DISPATCH-REPAINT instead of STREAM-REPLAY ensures, that we do not
+  ;; redraw everything twice when window-repaint-event happens, thanks
+  ;; to the event compression in the event queue. -- jd 2020-01-15
+  (dispatch-repaint pane (sheet-region pane)))
```
[the remark this answers is not in the document]
Comment (dkochmanski) [reply Core/clim-core/panes.lisp]: Previously we did draw and record at the same time. That lead to some problems[1] I've created a ticket with proposal[2] to address this kind of issues for recorded streams.

Basically at first we were drawing `+everywhere+` (so if a drawing operation did go beyond 16bit coordinate limit we had error from clx), and then repaint was limited to the viewport.

Now we repaint only records which are in the viewport (it is still possible that we will repaint it twice[3]).

[1] https://github.com/McCLIM/McCLIM/issues/793
[2] https://github.com/McCLIM/McCLIM/issues/927
[3] https://github.com/McCLIM/McCLIM/issues/937

File: Core/clim-core/frames.lisp (hunk 2 above)
Comment (dkochmanski) [reply Core/clim-core/frames.lisp]: the lookup in frame-panes-for-layout is because there because we create the interactor only once (in the `unless` form), `make-clim-interactor-pane` returns the wrapper pane as the first value (stream is returned as the second value).

**Comment by dkochmanski** [comment]
I've rebased the PR and answered comments. There were no code changes in the last push.

File: Core/clim-core/frames.lisp (hunk 1 above)
Comment (scymtym) [reply Core/clim-core/frames.lisp]: It is clear that `define-application-frame` has to generate code that binds `*application-frame*`, my question was whether `with-look-and-feel-realization` should do it.

The argument regarding the parameter makes sense to me.

**Comment by scymtym** [comment]
Not sure which commit caused this, but the "List Test" demo is broken. Using the scroll bar to scroll the list has no immediate effect while the pointer button is still pressed (there is some flickering in the scroll bar, but the viewport is not updated). When the button is released, the viewport updates.

**Comment by scymtym** [comment]
I don't know whether this is some other regression, but "Text size test" does not draw the "Family", "Face" and "Rectangle" gadgets and also flickers continuously. It is possibly that some or all of these problems are due to lack of rebasing onto master.

File: Core/clim-core/frames.lisp [hunk 4]
```diff
@@ -923,8 +935,9 @@ documentation produced by presentations.")
           ,@user-default-initargs)
          ,@other-options)
 
-       ,(generate-generate-panes-form
-         name menu-bar panes layouts pointer-documentation)
+       ,@(when panes
```
Comment (scymtym) [inline Core/clim-core/frames.lisp]: This should probably be `when (or panes layouts)` otherwise frames like `demodemo` that use layouts but do not use the `:panes` option don't work.

**Comment by dkochmanski** [comment]
missing gadgets indeed appeared after rebasing. flicker is because going through the queue is much slower. I'm withdrawing the display invocation change because there are other regressions related to Drei. I've cherry-picked fix for german towns demo regression.

as of display invocation and double initial repaint I'll address it sometime else if I'm able to overcome drei issue (it basically shortcuts repaint with redisplay).

File: Core/clim-core/frames.lisp (hunk 4 above)
Comment (dkochmanski) [reply Core/clim-core/frames.lisp]: fixed

**Comment by dkochmanski** [comment]
as of list scrolling issue it was due to dispatch-repaint (we should have used stream-replay instead, then the problem disappears), that said there are still problems with drei when doing that.

File: Backends/CLX/fonts.lisp [hunk 5]
```diff
@@ -41,7 +41,7 @@
     (when fonts
       (xlib:open-font display (first fonts)))))
 
-(defmethod text-style-mapping ((port clx-port) text-style &optional character-set
+(defmethod text-style-mapping ((port clx-port) (text-style text-style) &optional character-set
                                &aux (text-style (climb:parse-text-style* text-style)))
```
Comment (scymtym) [inline Backends/CLX/fonts.lisp]: Isn't `parse-text-style*` the identity when `text-style` already is a `text-style` instance?

Comment (dkochmanski) [reply Backends/CLX/fonts.lisp]: no, text-style may be incomplete (i.e it may have NIL components). It would be the identity for `parse-text-style` (without a star)

File: Core/clim-basic/drawing/medium.lisp [hunk 6]
```diff
@@ -295,8 +300,9 @@
 (defun parse-text-style (style)
   (cond ((text-style-p style) style)
         ((null style) (make-text-style nil nil nil)) ; ?
-        ((and (listp style) (<= 3 (length style) 4))
-         (apply #'make-text-style style))
+        ((and (listp style) (alexandria:length= 3 style))
+         (destructuring-bind (family face size) style
+           (make-text-style family face size)))
```
Comment (scymtym) [inline Core/clim-basic/drawing/medium.lisp]: I think this could be a separate commit which should also add tests for `parse-text-style`.

File: Backends/CLX/fonts.lisp (hunk 5 above)
Comment (scymtym) [reply Backends/CLX/fonts.lisp]: Right, I already forgot. But that reminds me: can we rename `parse-text-style*` to `ensure-fully-specified-text-style` or similar?

File: Core/clim-basic/drawing/medium.lisp (hunk 6 above)
Comment (dkochmanski) [reply Core/clim-basic/drawing/medium.lisp]: fixed

File: Backends/CLX/fonts.lisp (hunk 5 above)
Comment (dkochmanski) [reply Backends/CLX/fonts.lisp]: I'm not much into `overly-long-and-descriptive-name`'s, also it is indeed a parse-text-style variant, so "starred" name makes perfect sense here to me. note, that it is not only to ensure that the text style is fully specified but also to normalize the size

Comment (scymtym) [reply Backends/CLX/fonts.lisp]: Fair enough. I might bring it up again later, though :)
